Constantly keeping 10 idle connections

During investigation of https://github.com/r2dbc/r2dbc-pool/issues/115 I've noticed that there are idle connections which are never closed. I've created a repository which can be used to reproduce the issue https://github.com/petromir/r2dbc-pool-idle-connection-issue. Note that even if the maximum size of the pool is 2, we will always have 10 open idle connections to the database.

## Expected Behavior
When the application doesn't have any activity, it shouldn't use database resources.

## Actual Behavior
We always have 10 open idle connections to the database with initial query of `SELECT oid, typname FROM pg_catalog.pg_type WHERE typname IN ('hstore')`

## Steps to Reproduce
Use https://github.com/petromir/r2dbc-pool-idle-connection-issue

## Possible Solution
Provide an option to set the amount of connection kept in the background, e.g. `maxIdleConnections`.

## Your Environment

* r2dbc-pool - 0.9.0.M1
* r2dbc-postgresql - 0.8.7.RELEASE
* reactor-pool - 0.2.2
* reactor-core - 3.4.3
* Database: PostgreSQL 11.4
* JVM version: 1.8.0.202
* OS and version: macOS Big Sur 11.3 / Google Cloud Plaform

Well using background eviction doesn't help to close the actual connection, just making pool statistics look right.

The issue is easy to reproduce:
1. Clone my version of [r2dbc-pool](https://github.com/petromir/r2dbc-pool) with added support for background eviction. 
2. Run `mvn clean install -DskipTests`
3. Clone https://github.com/petromir/r2dbc-pool-idle-connection-issue and follow the steps defined in the repository.

As a result we have JMX statistics looking good (with 0 idle connection), BUT still 10 database connections in idle state.
```sql
select state, count(pid), query FROM pg_stat_activity where datname = 'books' and pid<>pg_backend_pid() and application_name = 'r2dbc-postgresql' group by state, query;
```
|state|count|query|
|-----|-----|-----|
|idle|10|INSERT INTO book (title, author) VALUES ($1, $2) RETURNING *|

Am I missing something?  

I've noticed that there is `.releaseHandler(Connection::close)` which results in two exceptions

- `Caused by: java.lang.IllegalStateException: The connection is closed`
This seems to be `r2dbc-pool` specific as we have assertion whether `io.r2dbc.pool.PooledConnection` is closed in several places
```java
private void assertNotClosed() {
        if (this.closed) {
            throw new IllegalStateException("The connection is closed");
        }
}
```
- Inability to acquire a connection.
```java
java.util.concurrent.TimeoutException: Did not observe any item or terminal signal within 7000ms in 'Connection Acquisition from [ConnectionPool[PostgreSQL]]' (and no fallback has been configured)
```

--- a/src/main/java/reactor/pool/AbstractPool.java
+++ b/src/main/java/reactor/pool/AbstractPool.java
@@ -166,8 +166,14 @@ abstract class AbstractPool<POOLABLE> implements InstrumentedPool<POOLABLE>,
 			});
 		}
 		else {
-			return Mono.from(factory.apply(poolable))
-					   .doFinally(fin -> metricsRecorder.recordDestroyLatency(clock.millis() - start));
+			Mono<Void> userProvidedDestroy;
+			try {
+				userProvidedDestroy = Mono.from(factory.apply(poolable));
+			}
+			catch (Throwable destroyFunctionError) {
+				userProvidedDestroy = Mono.error(destroyFunctionError);
+			}
+			return userProvidedDestroy.doFinally(fin -> metricsRecorder.recordDestroyLatency(clock.millis() - start));
 		}
 	}
 

--- a/src/main/java/reactor/pool/SimpleDequePool.java
+++ b/src/main/java/reactor/pool/SimpleDequePool.java
@@ -33,6 +33,7 @@ import org.reactivestreams.Subscription;
 import reactor.core.CoreSubscriber;
 import reactor.core.Disposable;
 import reactor.core.Disposables;
+import reactor.core.Exceptions;
 import reactor.core.Scannable;
 import reactor.core.publisher.Flux;
 import reactor.core.publisher.Mono;
@@ -148,7 +149,8 @@ public class SimpleDequePool<POOLABLE> extends AbstractPool<POOLABLE> {
 						if (pooledRef.markInvalidate()) {
 							recordInteractionTimestamp();
 							iterator.remove();
-							destroyPoolable(pooledRef).subscribe();
+							destroyPoolable(pooledRef).subscribe(v -> {},
+									destroyError -> this.logger.warn("Error while destroying resource in background eviction:", destroyError));
 						}
 					}
 				}
@@ -478,7 +480,7 @@ public class SimpleDequePool<POOLABLE> extends AbstractPool<POOLABLE> {
 	}
 
 	@SuppressWarnings("WeakerAccess")
-	final void maybeRecycleAndDrain(QueuePooledRef<POOLABLE> poolSlot) {
+	final void maybeRecycleAndDrain(QueuePooledRef<POOLABLE> poolSlot, CoreSubscriber<? super Void> actual) {
 		if (!isDisposed()) {
 			recordInteractionTimestamp();
 			if (!poolConfig.evictionPredicate()
@@ -489,6 +491,7 @@ public class SimpleDequePool<POOLABLE> extends AbstractPool<POOLABLE> {
 				if (irq != null) {
 					QueuePooledRef<POOLABLE> slot = recycleSlot(poolSlot);
 					irq.offerLast(slot);
+					actual.onComplete();
 					drain();
 					if (isDisposed() && slot.markInvalidate()) {
 						destroyPoolable(slot).subscribe(); //TODO manage errors?
@@ -497,10 +500,21 @@ public class SimpleDequePool<POOLABLE> extends AbstractPool<POOLABLE> {
 				}
 			}
 		}
+		//if we destroy instead of recycle, be sure to propagate destroy handler errors
+		//to the release() subscriber
 		if (poolSlot.markInvalidate()) {
 			destroyPoolable(poolSlot).subscribe(null,
-					e -> drain(),
-					this::drain); //TODO manage errors?
+					e -> {
+						actual.onError(e);
+						this.drain();
+					},
+					() -> {
+						actual.onComplete();
+						this.drain();
+					}); //TODO manage errors?
+		}
+		else {
+			actual.onComplete(); //already invalidated
 		}
 	}
 
@@ -591,7 +605,8 @@ public class SimpleDequePool<POOLABLE> extends AbstractPool<POOLABLE> {
 					//immediately clean up state
 					ACQUIRED.decrementAndGet(pool);
 					return pool.destroyPoolable(this)
-					           .then(Mono.fromRunnable(pool::drain));
+							//when invalidating, be sure to trigger drain even if there was an error
+							.doFinally(st -> pool.drain());
 				}
 				else {
 					return Mono.empty();
@@ -621,12 +636,18 @@ public class SimpleDequePool<POOLABLE> extends AbstractPool<POOLABLE> {
 					cleaner = pool.poolConfig.releaseHandler()
 					                         .apply(poolable);
 				}
-				catch (Throwable e) {
+				catch (Throwable releaseError) {
 					ACQUIRED.decrementAndGet(pool); //immediately clean up state
-					markReleased();
-					return Mono.error(new IllegalStateException(
-							"Couldn't apply cleaner function",
-							e));
+					if (markInvalidate()) {
+						return pool.destroyPoolable(this)
+								.onErrorResume(destroyError ->
+									Mono.error(new IllegalStateException("Couldn't apply releaseHandler nor destroyHandler",
+											Exceptions.multiple(releaseError, destroyError)))
+								)
+								.then(Mono.<Void>error(new IllegalStateException("Couldn't apply releaseHandler, resource destroyed", releaseError)))
+								.doFinally(st -> pool.drain());
+					}
+					return Mono.error(new IllegalStateException("Couldn't apply releaseHandler", releaseError)); //should be a very rare corner case
 				}
 				//the PoolRecyclerMono will wrap the cleaning Mono returned by the Function and perform state updates
 				return new QueuePoolRecyclerMono<>(cleaner, this);
@@ -695,8 +716,7 @@ public class SimpleDequePool<POOLABLE> extends AbstractPool<POOLABLE> {
 
 			pool.metricsRecorder.recordResetLatency(pool.clock.millis() - start);
 
-			pool.maybeRecycleAndDrain(slot);
-			actual.onComplete();
+			pool.maybeRecycleAndDrain(slot, actual);
 		}
 
 		@Override
@@ -733,9 +753,9 @@ public class SimpleDequePool<POOLABLE> extends AbstractPool<POOLABLE> {
 		@Override
 		public void onSubscribe(Subscription s) {
 			if (Operators.validate(upstream, s)) {
+				this.start = pool.clock.millis();
 				this.upstream = s;
 				actual.onSubscribe(this);
-				this.start = pool.clock.millis();
 			}
 		}
 

--- a/src/test/java/reactor/pool/CommonPoolTest.java
+++ b/src/test/java/reactor/pool/CommonPoolTest.java
@@ -2599,6 +2599,7 @@ public class CommonPoolTest {
 
 		//assert that next acquire will work
 		PooledRef<String> ref3 = pool.acquire().block(Duration.ofSeconds(1));
+		assert ref3 != null;
 		assertThat(ref3.poolable()).as("poolable 3").isEqualTo("releaseHandlerCase3");
 	}
 
@@ -2632,6 +2633,7 @@ public class CommonPoolTest {
 
 		//assert that next acquire will work
 		PooledRef<String> ref3 = pool.acquire().block(Duration.ofSeconds(1));
+		assert ref3 != null;
 		assertThat(ref3.poolable()).as("poolable 3").isEqualTo("destroyHandlerInvalidateCase3");
 	}
 
@@ -2666,6 +2668,7 @@ public class CommonPoolTest {
 
 		//assert that next acquire will work
 		PooledRef<String> ref3 = pool.acquire().block(Duration.ofSeconds(1));
+		assert ref3 != null;
 		assertThat(ref3.poolable()).as("poolable 3").isEqualTo("destroyHandlerReleaseCase3");
 	}
 
@@ -2686,7 +2689,9 @@ public class CommonPoolTest {
 		InstrumentedPool<String> pool = style.apply(configBuilder);
 
 		//generate a first value and immediately release it back to the pool
-		pool.acquire().block(Duration.ofMillis(100)).release().block(Duration.ofMillis(500));
+		PooledRef<String> initialRef = pool.acquire().block(Duration.ofMillis(100));
+		assert initialRef != null;
+		initialRef.release().block(Duration.ofMillis(500));
 
 		//trigger background eviction
 		assertThatCode(() -> vts.advanceTimeBy(Duration.ofSeconds(10))).doesNotThrowAnyException();
@@ -2694,10 +2699,9 @@ public class CommonPoolTest {
 
 		//assert that next acquire will work
 		PooledRef<String> ref2 = pool.acquire().block(Duration.ofSeconds(1));
+		assert ref2 != null;
 		assertThat(ref2.poolable()).as("poolable 2").isEqualTo("destroyHandlerBackgroundCase2");
 	}
-
-
 
 	@ParameterizedTestWithName
 	@EnumSource
@@ -2736,8 +2740,7 @@ public class CommonPoolTest {
 
 		//assert that next acquire will work
 		PooledRef<String> ref3 = pool.acquire().block(Duration.ofSeconds(1));
+		assert ref3 != null;
 		assertThat(ref3.poolable()).as("poolable 3").isEqualTo("bothReleaseHandlerAndDestroyHandlerFail3");
 	}
-
-
 }

--- a/src/test/java/reactor/pool/CommonPoolTest.java
+++ b/src/test/java/reactor/pool/CommonPoolTest.java
@@ -48,6 +48,7 @@ import org.junit.jupiter.params.provider.MethodSource;
 
 import reactor.core.Disposable;
 import reactor.core.Disposables;
+import reactor.core.Exceptions;
 import reactor.core.publisher.Flux;
 import reactor.core.publisher.Mono;
 import reactor.core.publisher.SignalType;
@@ -56,6 +57,7 @@ import reactor.core.scheduler.Schedulers;
 import reactor.pool.InstrumentedPool.PoolMetrics;
 import reactor.pool.TestUtils.ParameterizedTestWithName;
 import reactor.pool.TestUtils.PoolableTest;
+import reactor.scheduler.clock.SchedulerClock;
 import reactor.test.StepVerifier;
 import reactor.test.publisher.TestPublisher;
 import reactor.test.scheduler.VirtualTimeScheduler;
@@ -1784,9 +1786,6 @@ public class CommonPoolTest {
 		pool.acquire().flatMap(PooledRef::release).block();
 		pool.acquire().flatMap(PooledRef::release).block();
 
-		//destroy is fire-and-forget so the 500ms one will not have finished
-		assertThat(recorder.getDestroyCount()).as("destroy before 500ms").isEqualTo(1);
-
 		await().pollDelay(500, TimeUnit.MILLISECONDS)
 		       .atMost(1, TimeUnit.SECONDS)
 		       .untilAsserted(() -> assertThat(recorder.getDestroyCount()).as("destroy after 500ms").isEqualTo(2));
@@ -2559,4 +2558,186 @@ public class CommonPoolTest {
 
 		assertThat(pool.metrics().acquiredSize()).as("acquired size after allocator done").isOne();
 	}
+
+	private static Mono<Void> throwingFunction(String poolable) {
+		if (poolable.endsWith("1")) {
+			IllegalStateException e = new IllegalStateException("(expected) throwing instead of producing a Mono for " + poolable);
+			StackTraceElement[] traceElements = e.getStackTrace();
+			e.setStackTrace(Arrays.copyOf(traceElements, 5)); //don't fill the logs with too much stack trace
+			throw e;
+		}
+		return Mono.empty();
+	}
+
+	@ParameterizedTestWithName
+	@EnumSource
+	void releaseHandlerProtectedAgainstThrowingFunction(PoolStyle style) {
+		AtomicInteger count = new AtomicInteger();
+		PoolBuilder<String, PoolConfig<String>> configBuilder = PoolBuilder
+				.from(Mono.fromCallable(() -> "releaseHandlerCase" + count.incrementAndGet()))
+				.sizeBetween(0, 1)
+				.releaseHandler(CommonPoolTest::throwingFunction);
+
+		InstrumentedPool<String> pool = style.apply(configBuilder);
+
+		PooledRef<String> ref1 = pool.acquire().block();
+		assert ref1 != null;
+
+		//set up a concurrent acquire (blocked for now)
+		AtomicReference<PooledRef<String>> acquire2 = new AtomicReference<>();
+		pool.acquire().subscribe(acquire2::set);
+
+		StepVerifier.create(ref1.release())
+				.expectErrorSatisfies(t -> assertThat(t)
+						.hasMessage("Couldn't apply releaseHandler, resource destroyed")
+						.hasCause(new IllegalStateException("(expected) throwing instead of producing a Mono for releaseHandlerCase1")))
+				.verify(Duration.ofSeconds(1));
+
+		//assert that concurrent acquire has been unstuck, then destroy it
+		assertThat(acquire2.get()).as("concurrent acquire2").isNotNull();
+		acquire2.get().invalidate().block();
+
+		//assert that next acquire will work
+		PooledRef<String> ref3 = pool.acquire().block(Duration.ofSeconds(1));
+		assertThat(ref3.poolable()).as("poolable 3").isEqualTo("releaseHandlerCase3");
+	}
+
+	@ParameterizedTestWithName
+	@EnumSource
+	void destroyHandlerProtectedAgainstThrowingFunction_invalidate(PoolStyle style) {
+		AtomicInteger count = new AtomicInteger();
+		PoolBuilder<String, PoolConfig<String>> configBuilder = PoolBuilder
+				.from(Mono.fromCallable(() -> "destroyHandlerInvalidateCase" + count.incrementAndGet()))
+				.sizeBetween(0, 1)
+				.destroyHandler(CommonPoolTest::throwingFunction);
+
+		InstrumentedPool<String> pool = style.apply(configBuilder);
+
+		PooledRef<String> ref1 = pool.acquire().block();
+		assert ref1 != null;
+
+		//set up a concurrent acquire (blocked for now)
+		AtomicReference<PooledRef<String>> acquire2 = new AtomicReference<>();
+		pool.acquire().subscribe(acquire2::set);
+
+		StepVerifier.create(ref1.invalidate())
+				.expectErrorSatisfies(t -> assertThat(t)
+						.hasMessage("(expected) throwing instead of producing a Mono for destroyHandlerInvalidateCase1")
+						.hasNoCause())
+				.verify(Duration.ofSeconds(1));
+
+		//assert that concurrent acquire has been unstuck, then destroy it
+		assertThat(acquire2.get()).as("concurrent acquire2").isNotNull();
+		acquire2.get().invalidate().block();
+
+		//assert that next acquire will work
+		PooledRef<String> ref3 = pool.acquire().block(Duration.ofSeconds(1));
+		assertThat(ref3.poolable()).as("poolable 3").isEqualTo("destroyHandlerInvalidateCase3");
+	}
+
+	@ParameterizedTestWithName
+	@EnumSource
+	void destroyHandlerProtectedAgainstThrowingFunction_releaseWithEviction(PoolStyle style) {
+		AtomicInteger count = new AtomicInteger();
+		PoolBuilder<String, PoolConfig<String>> configBuilder = PoolBuilder
+				.from(Mono.fromCallable(() -> "destroyHandlerReleaseCase" + count.incrementAndGet()))
+				.sizeBetween(0, 1)
+				.evictionPredicate((s, ref) -> ref.acquireCount() == 1 && s.endsWith("1"))
+				.destroyHandler(CommonPoolTest::throwingFunction);
+
+		InstrumentedPool<String> pool = style.apply(configBuilder);
+
+		PooledRef<String> ref1 = pool.acquire().block();
+		assert ref1 != null;
+
+		//set up a concurrent acquire (blocked for now)
+		AtomicReference<PooledRef<String>> acquire2 = new AtomicReference<>();
+		pool.acquire().subscribe(acquire2::set, Throwable::printStackTrace);
+
+		StepVerifier.create(ref1.release())
+				.expectErrorSatisfies(t -> assertThat(t)
+						.hasMessage("(expected) throwing instead of producing a Mono for destroyHandlerReleaseCase1")
+						.hasNoCause())
+				.verify(Duration.ofSeconds(1));
+
+		//assert that concurrent acquire has been unstuck, then destroy it
+		assertThat(acquire2.get()).as("concurrent acquire2").isNotNull();
+		acquire2.get().invalidate().block();
+
+		//assert that next acquire will work
+		PooledRef<String> ref3 = pool.acquire().block(Duration.ofSeconds(1));
+		assertThat(ref3.poolable()).as("poolable 3").isEqualTo("destroyHandlerReleaseCase3");
+	}
+
+	@ParameterizedTestWithName
+	@EnumSource
+	void destroyHandlerProtectedAgainstThrowingFunction_evictInBackground(PoolStyle style) {
+		VirtualTimeScheduler vts = VirtualTimeScheduler.create();
+		AtomicInteger count = new AtomicInteger();
+		PoolBuilder<String, PoolConfig<String>> configBuilder = PoolBuilder
+				.from(Mono.fromCallable(() -> "destroyHandlerBackgroundCase" + count.incrementAndGet()))
+				.sizeBetween(0, 1)
+				.clock(SchedulerClock.of(vts))
+				.evictInBackground(Duration.ofSeconds(4), vts)
+				//we need to assert lifetime, which is precise thanks to VTS, otherwise acquire().release() would trip eviction
+				.evictionPredicate((s, ref) -> ref.acquireCount() == 1 && ref.lifeTime() >= 4000)
+				.destroyHandler(CommonPoolTest::throwingFunction);
+
+		InstrumentedPool<String> pool = style.apply(configBuilder);
+
+		//generate a first value and immediately release it back to the pool
+		pool.acquire().block(Duration.ofMillis(100)).release().block(Duration.ofMillis(500));
+
+		//trigger background eviction
+		assertThatCode(() -> vts.advanceTimeBy(Duration.ofSeconds(10))).doesNotThrowAnyException();
+		assertThat(pool.metrics().idleSize()).as("background eviction ran").isZero();
+
+		//assert that next acquire will work
+		PooledRef<String> ref2 = pool.acquire().block(Duration.ofSeconds(1));
+		assertThat(ref2.poolable()).as("poolable 2").isEqualTo("destroyHandlerBackgroundCase2");
+	}
+
+
+
+	@ParameterizedTestWithName
+	@EnumSource
+	void releaseHandlerFailsThenDestroyHandlerAlsoFails(PoolStyle style) {
+		AtomicInteger count = new AtomicInteger();
+		PoolBuilder<String, PoolConfig<String>> configBuilder = PoolBuilder
+				.from(Mono.fromCallable(() -> "bothReleaseHandlerAndDestroyHandlerFail" + count.incrementAndGet()))
+				.sizeBetween(0, 1)
+				.evictionPredicate((s, ref) -> ref.acquireCount() == 1)
+				.releaseHandler(CommonPoolTest::throwingFunction)
+				.destroyHandler(CommonPoolTest::throwingFunction);
+
+		InstrumentedPool<String> pool = style.apply(configBuilder);
+
+		PooledRef<String> ref1 = pool.acquire().block();
+		assert ref1 != null;
+
+		//set up a concurrent acquire (blocked for now)
+		AtomicReference<PooledRef<String>> acquire2 = new AtomicReference<>();
+		pool.acquire().subscribe(acquire2::set, Throwable::printStackTrace);
+
+		StepVerifier.create(ref1.release())
+				.expectErrorSatisfies(t -> {
+					assertThat(t)
+							.isInstanceOf(IllegalStateException.class)
+							.hasMessage("Couldn't apply releaseHandler nor destroyHandler");
+					assertThat(Exceptions.unwrapMultiple(t.getCause()).stream().map(Throwable::getMessage))
+							.containsExactly("(expected) throwing instead of producing a Mono for bothReleaseHandlerAndDestroyHandlerFail1",
+									"(expected) throwing instead of producing a Mono for bothReleaseHandlerAndDestroyHandlerFail1");
+				})
+				.verify(Duration.ofSeconds(1));
+
+		//assert that concurrent acquire has been unstuck, then destroy it
+		assertThat(acquire2.get()).as("concurrent acquire2").isNotNull();
+		acquire2.get().invalidate().block();
+
+		//assert that next acquire will work
+		PooledRef<String> ref3 = pool.acquire().block(Duration.ofSeconds(1));
+		assertThat(ref3.poolable()).as("poolable 3").isEqualTo("bothReleaseHandlerAndDestroyHandlerFail3");
+	}
+
+
 }

